kufar/widget/main.go

```diff
@@ -13,32 +13,48 @@ type widget struct {
 	time  time.Time
 }
 
-var n = flag.Int("n", 1, "how many widgets produced by producer")
+var wid = flag.Int("w", 1, "how many widgets produced by producer")
+var con = flag.Int("c", 1, "number of consumers")
+var dur = flag.Duration("d", time.Second, "a consumer taking a while to process a widget")
 
 func main() {
 	c := make(chan widget)
-	var wg sync.WaitGroup
+	var cwg sync.WaitGroup
+	var pwg sync.WaitGroup
 	flag.Parse()
 
-	for i := 1; i <= *n; i++ {
-		wg.Add(1)
+	cwg.Add(*con)
+	for i := 0; i <= *con; i++ {
+		go consumer(c, i, &cwg)
+	}
+
+	ticker := time.NewTicker(*dur)
+	var tickCounter int
+
+	pwg.Add(*wid)
+	for range ticker.C {
+		if tickCounter >= *wid {
+			ticker.Stop()
+			break
+		}
+
 		go func(num int) {
-			defer wg.Done()
 			c <- widget{label: "widget_" + strconv.Itoa(num), time: time.Now()}
-		}(i)
+			pwg.Done()
+		}(tickCounter)
+
+		tickCounter++
 	}
 
-	go func(){
-		wg.Wait()
+	go func() {
```

Comment:
Please remove that go func.

IN your main func it need to be `close(c)` and after that `cwg.Wait` to wait until all consumers will finish with processing